Jetpack sign in - Step 1: Fetch Jetpack e-mail address for flowing through Magic Links

anitaa1990 opened the pull request:
Partially fixes #1482. This PR completes the first step of the Jetpack sign in with self credentials as defined in the master thread. 

### Changes
- ~~Revert existing logic of validating site url from `FETCH_CONNECT_SITE_INFO` endpoint and use the `FETCH_WPCOM_SITE_BY_URL` endpoint and initiate the discovery process.~~ Since using the discovery endpoint to validate the site would interfere with the existing logic of logging in using wp.com credentials, I've moved the discovery process to initiate only when the user enters their username + password in the `LoginUsernamePasswordFragment`.
- Added a new button to `LoginEmailFragment`: `Login via site credentials`.
- Design changes:
  - Added label to the top of the `LoginUsernamePasswordFragment`.
  - Change hint text of `username` in `LoginUsernamePasswordFragment`.
- When site credentials are clicked:
  - Initialise the discovery process which will let us know if the XMLRPC endpoint is supported by the site. **Note that the error handling during the discovery process will take place in another PR.**
   - If jetpack is installed/active/connected: call the `wp.getOptions` endpoint to fetch the `jetpack_user_email` and store in `SiteModel` and redirect to the request magic link screen.
   - If jetpack NOT installed/active/connected: Redirect to a new fragment called `LoginNoJetpackFragment`. 


### Notes:
- This PR handles only the redirection to the magic link login. The design changes required in the magic link login will take place in another PR.
- Since this PR adds logic to initiate the discovery process, there are new error messages and error scenarios that need to be incorporated. This will likely take place in another PR.
- ~Since this PR removes the logic to validate site address using the `FETCH_CONNECT_SITE_INFO` endpoint, there are some events that will likely be invalidated once this feature is merged.~
   - `LOGIN_CONNECTED_SITE_INFO_FAILED`
   - `LOGIN_CONNECTED_SITE_INFO_SUCCEEDED` 
- The are TODO's added to the PR to add events to the button clicks. This will take place in another PR.
- Note that this PR is to be merged to a feature branch. 


### Deviations from design:
- The username hint in `LoginUsernamePasswordFragment` is set to just `Username` rather than just `Username or Email` because there is no option to pass email address via the API, just the username.

### Behaviour
(LEFT: Login with self hosted site with Jetpack . RIGHT: Login with self hosted site without Jetpack)
<img width="300" src="https://user-images.githubusercontent.com/22608780/67195788-3c9f2a80-f417-11e9-97ba-7fb73e6e826a.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/67195794-3f018480-f417-11e9-8897-8afac854b4a2.gif">

##### Entering a non existent site:
<img width="300" src="https://user-images.githubusercontent.com/22608780/67195997-a4ee0c00-f417-11e9-9e75-c3da2669b6ad.gif">

#### Testing
- Verify that existing flow works as expected.
- Enter a non existent site and verify that an error message is displayed.
- Enter a non jetpack site and verify that Jetpack required screen is displayed.
  - Click on `Sign in` at the bottom of the screen.
  - Verify that the Email screen is displayed. Click on `Login with site credentials`.
  - Enter username + password and click on `Next`.
  - Verify that you are redirected to Jetpack needed screen.
  - Install, activate and connect Jetpack on your test site while in the app.
  - Click on `Refresh the app` link. Notice that you are redirected to the Site credentials screen.
  - Click on `Next` and verify that you are redirected to the Magic link sign in screen.
- Enter a jetpack site and verify that you are redirect to the email screen.
  - Click on `Login with site credentials`.
  - Enter username + password and click on `Next` and verify that you are redirected to the Magic link sign in screen.
  

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

anitaa1990 committed:
Display login with site credentials link button in login email fragment

anitaa1990 committed:
Added label to the top of the LoginUsernamePasswordFragment for Woo login only

anitaa1990 committed:
Revert the connection info endpoint implementation and update error message when site is not found

anitaa1990 committed:
Added logic to redirect Woo users to email login or jetpack required screen

anitaa1990 committed:
Initiate discovery process even if jetpack is not installed/active/connected

anitaa1990 committed:
Pass site xmlrpc url to email and jetpack required screen

anitaa1990 committed:
Added logic to fetch jetpack user email from site credentials and redirect to magic link screen

anitaa1990 committed:
Change hint text of username

anitaa1990 committed:
Add separate UI to display no jetpack screen

anitaa1990 committed:
Revert adding a new method to LoginListener, instead use an existing listener method

anitaa1990 committed:
Added logic to redirect to no Jetpack screen when site credentials are entered

anitaa1990 committed:
Added logic to fetch SiteModel based on the site url

AmandaRiu reviewed (commented):
@anitaa1990 Just a few review notes from tonight that are not related to our conversation around the discovery process that will change a bunch of this PR:

* Keyboard stays visible when navigating to the “Magic Link” screen as well as the new `LoginNoJetpackFragment`, but there are no text fields in these views.

<img src="https://user-images.githubusercontent.com/5810477/67259318-08912d00-f44a-11e9-89ca-eb78065974cf.png" width="300"/>

* The new `LoginNoJetpackFragment`  was a bit confusing since we already have a screen almost identical to this one. Is this necessary because technically the user isn’t logged into WPcom when they view it? It would be helpful if the layout file was populated with values similar to what the `LoginNoJetpackFragment` would populate it with so it’s easier to visualize since it’s only used by this fragment. It would also be helpful to remove anything that isn’t used by this fragment like the `user_info_group` and renaming the root from `site_picker_root` as it leads to a lot of confusion when trying to read the code. 
* Also regarding the `LoginNoJetpackFragment`: 

<img src="https://user-images.githubusercontent.com/5810477/67259370-3f674300-f44a-11e9-9348-7d9522808c20.png" width="300"/>

 Where would I find this screen in the designs? The **refresh the app to continue** and **Try again** buttons do the same thing. The **Log out** in the username message looks like it should do something but it does not. I wonder if it would make more sense to have it follow the same layout as the current similar screen:

<img src="https://user-images.githubusercontent.com/5810477/67259384-586ff400-f44a-11e9-9448-85afd220d924.png" width="300"/>

...except we change the primary button to **LOGOUT** (instead of having it at the top of the screen), and then set the secondary button to **VIEW SETUP INSTRUCTIONS**. Just some ideas if there is no official design yet. Looking forward to the next round!

anitaa1990 committed:
Revert discovery initiation changes to the site fragment

anitaa1990 committed:
Added separate method to redirect to username screen in LoginListener

anitaa1990 committed:
Added logic to pass the input site address to forgot password flow, if the xmlrpc url is empty

anitaa1990 committed:
Added logic to initiate discovery process for xmlrpc endpoint

anitaa1990 committed:
Added click listener to logout button and made design modificiations to LoginNoJetpackFragment

anitaa1990 committed:
Hide keyboard before redirecting to the magic link/no jetpack screen

anitaa1990 commented:
Thanks for the review @AmandaRiu! 

As discussed, I've reverted to logic to validate the site using the discovery API in e601f9b. The discovery process will now be initiated only when the user clicks on the `Login with site credentials` and enters their username + password. Since there is no way to complete the jetpack sign in flow implemented in this PR, without calling the discovery API, I've moved this logic to a separate class called `LoginBaseDiscoveryFragment`. This only applies to Woo at the moment. I have not modified the existing implementation used by wp-android in `LoginSiteAddressFragment`.

> Keyboard stays visible when navigating to the “Magic Link” screen as well as the new LoginNoJetpackFragment, but there are no text fields in these views.

I was not able to replicate this but I added code to close the keypad before redirecting to the "Magic link" or `LoginNoJetpackFragment` in aa28af6.

> The new LoginNoJetpackFragment was a bit confusing since we already have a screen almost identical to this one. Is this necessary because technically the user isn’t logged into WPcom when they view it? 

We need Jetpack to be installed/active/connected in order to support signing in with their site credentials. Only when Jetpack is connected, would we be able to get the `jetpack_user_email` connected with their username + password. And once the email is fetched, we can redirect them to the "Magic link" flow. I created a new fragment and xml file rather than reuse the existing one, because the existing one is in the `SitePickerActivity` which has a lot of logic in place to check if WooCommerce exists or if user is logged in etc. that is not relevant to this logic here. Since the layout is similar to what we use in the `SitePickerActivity` (minus the `RecyclerView` and some layouts specific to that screen), I just copied the layout relevant to the `LoginNoJetpackFragment` screen.

> It would also be helpful to remove anything that isn’t used by this fragment like the user_info_group and renaming the root from site_picker_root as it leads to a lot of confusion when trying to read the code.

Done in 40b0360

> The refresh the app to continue and Try again buttons do the same thing. 

Agreed! I removed the `Refresh the app to continue` in 40b0360 since it was not in the design and since the `Try Again` button does the same thing.

> The Log out in the username message looks like it should do something but it does not.

I forgot to add the logout click functionality before. I added it in 40b0360.

The design specs have two separate messages for when Jetpack is **NOT** installed and when Jetpack is installed but not connected. But since there is no way to know from the wp.getOptions API if jetpack is installed or installed but not connected, we're currently using only the LEFT side message in the app:

<img width="300" src="https://user-images.githubusercontent.com/22608780/67286985-e8b84280-f4f7-11e9-9b05-e46f5a99048a.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/67286783-9e36c600-f4f7-11e9-91a2-698e7531958a.png">

The entire design flow can be found in #1482 . PR ready for another round 👍

AmandaRiu commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -99,6 +100,7 @@
     <string name="cant_open_url">Unable to open the link</string>
     <string name="already_logged_in_wpcom">You\'re already logged in a WordPress.com account, you can\'t add a WordPress.com site bound to another account.</string>
     <string name="at_username">\@%s</string>
+    <string name="login_no_jetpack_username">Signed in as \@1%s\nWrong account? %2$s</string>
```
This is prepending a `1` to the username. Should be `Signed in as \@%1$s\nWrong account? %2$s`

anitaa1990 replied:
👌 Fixed in 4edd1cf

AmandaRiu reviewed (changes requested):
@anitaa1990  I've finished another round and really excited about the progress so far! 

## Testing

I ran the following tests (all passed):

#### Test Current Functionality
+ Basic login -  with Jetpack and Woo
+ Basic login - without Jetpack
+ Basic login -  without Woo
+ Basic login - site does not belong to WPcom account

#### Test New Functionality
+ Site creds login - with Jetpack and Woo
+ Site creds login - without Jetpack 
+ Enable jetpack from last step and refresh - should get redirected to the site creds screen.
+ Site creds login - without Woo
+ Site creds login - site with multiple jetpacked accounts

#### Other Testing
+ Log in to a non-existent site
+ Log in with a non-WP site
+ Enter a jetpack site and verify directed to the magic link screen
+ Enter a bad password for site creds
+ Enter an invalid username for site creds

#### NoJetpackFragment Testing
+ Logout functionality
+ Try again functionality
+ View instructions

## Code Review

I also left a bunch of code comments in addition to the following:

> This only applies to Woo at the moment. I have not modified the existing implementation used by wp-android in LoginSiteAddressFragment.

We should definitely port that change to the `LoginSiteAddressFragment` to avoid issues in the future. Since you’re adding this base class in this PR I think it makes the most sense to port that over in this PR to keep it altogether.

> I created a new fragment and xml file rather than reuse the existing one, because the existing one is in the SitePickerActivity which has a lot of logic in place to check if WooCommerce exists or if user is logged in etc. that is not relevant to this logic here. 

Makes sense to me! 👍

anitaa1990 committed:
Moved interface for LoginNoJetpackFragment to a separate class

anitaa1990 committed:
Fixed nitpicks in code/xml

anitaa1990 committed:
Handle corner case in discovery process and redirect to email screen if site is wp.com

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/LoginJetpackRequiredFragment.kt:
```diff
@@ -45,7 +45,13 @@ class LoginJetpackRequiredFragment : Fragment() {
+            endpointAddress: String?,
+            inputUsername: String?,
+            inputPassword: String?
+        )
```
Sounds good 👍 Fixed in 66eea8a

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/LoginNoJetpackFragment.kt:
```diff
@@ -0,0 +1,164 @@
+
+        // this will throw if parent activity doesn't implement the login listener interface
+        loginListener = context as? LoginListener
+        jetpackLoginListener = context as? LoginJetpackRequiredListener
```
Modified in both `LoginNoJetpackFragment` and `LoginJetpackRequiredFragment` in 66eea8a

anitaa1990 commented on libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/LoginUsernamePasswordFragment.java:
```diff
@@ -322,6 +343,28 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
+
+    @Override
+    public void handleWpComDiscoveryError(String failedEndpoint) {
+        // Used only by wp-android as Woo already uses the Connect-Inf-url to check
```
Makes sense. wp-android adds logic to redirect the user to the email screen so they can login using their wp.com email and password. I've added the same logic here in 6b94b05.

anitaa1990 commented on libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/LoginUsernamePasswordFragment.java:
```diff
@@ -510,6 +553,33 @@ public void onSiteChanged(OnSiteChanged event) {
+                                lastAddedXMLRPCSite.getXmlRpcUrl(), lastAddedXMLRPCSite.getUsername(),
+                                lastAddedXMLRPCSite.getPassword(), mAccountStore.getAccount().getAvatarUrl());
+                    } else {
+                        mLoginListener.gotWpcomEmail(lastAddedXMLRPCSite.getJetpackUserEmail());
```
Fixed in 4edd1cf

anitaa1990 commented on libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/LoginUsernamePasswordFragment.java:
```diff
@@ -215,7 +224,7 @@ public void onCreate(Bundle savedInstanceState) {
 
         mInputSiteAddress = getArguments().getString(ARG_INPUT_SITE_ADDRESS);
-        mEndpointAddress = getArguments().getString(ARG_ENDPOINT_ADDRESS);
+        mEndpointAddress = getArguments().getString(ARG_ENDPOINT_ADDRESS, "");
```
👍 Fixed in 4edd1cf

anitaa1990 commented on libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/LoginUsernamePasswordFragment.java:
```diff
@@ -144,11 +151,13 @@ protected void setupContent(ViewGroup rootView) {
 
-        mInputSiteAddressWithoutSuffix = UrlUtils.removeXmlrpcSuffix(mEndpointAddress);
+        String inputSiteAddressWithoutProtocol = UrlUtils.removeScheme(mInputSiteAddress);
+        mInputSiteAddressWithoutSuffix = mEndpointAddress.isEmpty()
```
👍 Done in 4edd1cf

anitaa1990 commented on libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/LoginBaseDiscoveryFragment.java:
```diff
@@ -0,0 +1,143 @@
+    private void askForHttpAuthCredentials(@NonNull final String url) {
+        LoginHttpAuthDialogFragment loginHttpAuthDialogFragment = LoginHttpAuthDialogFragment.newInstance(url);
+        loginHttpAuthDialogFragment.setTargetFragment(this, LoginHttpAuthDialogFragment.DO_HTTP_AUTH);
+        loginHttpAuthDialogFragment.show(getFragmentManager(), LoginHttpAuthDialogFragment.TAG);
```
😢 This is why I like kotlin. I've added a null check in 4edd1cf

anitaa1990 commented:
Thanks for the thorough review @AmandaRiu! I've fixed all the code changes and PR ready for another round 👍 

> We should definitely port that change to the LoginSiteAddressFragment to avoid issues in the future. Since you’re adding this base class in this PR I think it makes the most sense to port that over in this PR to keep it altogether.

I agree that we should definitely port the `LoginBaseDiscooveryFragment` changes to the `LoginSiteAddressFragment`. But I would like to work on this in another PR, if that's ok. This is because there is some discussion around how best to handle the error scenarios in our app:
![download](https://user-images.githubusercontent.com/22608780/67369254-51ee9300-f596-11e9-8ca4-af98bed11738.png)

And based on what we decide, we would be making changes to this class and it would make it easier to implement the changes first and then port things over from `LoginSiteAddressFragment`. What do you think? I've added a separate task for this in the master issue thread [here](https://github.com/woocommerce/woocommerce-android/issues/1482), so it does not get missed.

AmandaRiu reviewed (approved):
LGTM! :shipit:
